src/coreclr/jit/lsra.cpp

```diff
@@ -2309,12 +2308,43 @@ BasicBlock* LinearScan::findPredBlockForLiveIn(BasicBlock* block,
 
     if (block->bbPreds == nullptr)
     {
+        assert((block != compiler->fgFirstBB) || (prevBlock != nullptr));
+        JITDUMP("\n\nNo predecessor; ");
+
+        // The shared throw blocks do not have predecessor because no block's bbJumpDest points to
+        // them. For such blocks, we want to return the fact that predecessor is indeed null instead
+        // of returning prevBlock. Returning prevBlock will be wrong, because LSRA would think that
+        // the variable is live in registers based on the lexical flow, but that won't be true from
+        // control flow perspective.
+        // Example:
+        //
+        // IG05:
+        //      ...         ; V01 is in 'rdi'
+        //      JNE IG07
+        //      ...
+        // IG06:
+        //      ...
+        //      ...         ; V01 is in 'rbx'
+        //      JMP IG08
+        // IG07:
+        //      ...         ; LSRA thinks V01 is in 'rbx' if IG06 is set as previous block of IG07.
+        //      ....
+        //      CALL CORINFO_HELP_RNGCHKFAIL
+        //      ...
+        // IG08:
+        //      ...
+        //      ...
+        if (block->bbJumpKind == BBJ_THROW && ((block->bbFlags & BBF_SHARED_THROW) != 0))
+        {
+            assert((block->bbFlags & BBF_DONT_REMOVE) != 0);
+            JITDUMP(" - Shared throw block; ");
+            return nullptr;
+        }
+
         // We may have unreachable blocks, due to optimization.
         // We don't want to set the predecessor as null in this case, since that will result in
         // unnecessary DummyDefs, and possibly result in inconsistencies requiring resolution
         // (since these unreachable blocks can have reachable successors).
```

Comment:
Did you try just always returning null and look at diffs, to understand this comment? Maybe it wouldn't apply to any throw block since I believe there would never be a register resolution between a throw block and any successor? That is, maybe you can return null for any throw block with no predecessor, and not worry about it being a special throw helper block.

Reply:
I see slight regression if just return `nullptr` if there is no previous block.

```
Total PerfScoreUnits of base: 65337.50000000002
Total PerfScoreUnits of diff: 65373.940000000024
Total PerfScoreUnits of delta: 36.44 (0.06% of base)
    diff is a regression.

```
<details>
<summary>Details perfscore diff</summary>

```
Top file regressions (PerfScoreUnits):
       32.40 : 191062.dasm (0.63% of base)
        0.90 : 74188.dasm (0.27% of base)
        0.85 : 231849.dasm (0.20% of base)
        0.85 : 230436.dasm (0.20% of base)
        0.85 : 230451.dasm (0.19% of base)
        0.85 : 231864.dasm (0.19% of base)
        0.85 : 231040.dasm (0.20% of base)
        0.85 : 231055.dasm (0.19% of base)
        0.55 : 74042.dasm (0.04% of base)
        0.50 : 231835.dasm (0.12% of base)
        0.50 : 230422.dasm (0.12% of base)
        0.50 : 231026.dasm (0.12% of base)

Top file improvements (PerfScoreUnits):
       -0.90 : 193548.dasm (-0.40% of base)
       -0.75 : 106063.dasm (-0.20% of base)
       -0.48 : 193493.dasm (-0.59% of base)
       -0.47 : 193611.dasm (-0.54% of base)
       -0.47 : 193509.dasm (-0.20% of base)
       -0.47 : 193518.dasm (-0.21% of base)
       -0.47 : 193632.dasm (-0.32% of base)

19 total files with Perf Score differences (7 improved, 12 regressed), 34 unchanged.

Top method regressions (PerfScoreUnits):
       32.40 ( 0.63% of base) : 191062.dasm - System.Linq.Parallel.OrderPreservingPipeliningSpoolingTask`2[__Canon,Nullable`1][System.__Canon,System.Nullable`1[System.Int32]]:SpoolingWork():this
        0.90 ( 0.27% of base) : 74188.dasm - Microsoft.CodeAnalysis.AssemblyPortabilityPolicy:LoadFromXml(System.IO.Stream):Microsoft.CodeAnalysis.AssemblyPortabilityPolicy
        0.85 ( 0.20% of base) : 231849.dasm - Xunit.Sdk.AssertEqualityComparer`1[Int32][System.Int32]:Equals(int,int):bool:this
        0.85 ( 0.20% of base) : 230436.dasm - Xunit.Sdk.AssertEqualityComparer`1[Int32][System.Int32]:Equals(int,int):bool:this
        0.85 ( 0.19% of base) : 230451.dasm - Xunit.Sdk.AssertEqualityComparer`1[Int64][System.Int64]:Equals(long,long):bool:this
        0.85 ( 0.19% of base) : 231864.dasm - Xunit.Sdk.AssertEqualityComparer`1[Int64][System.Int64]:Equals(long,long):bool:this
        0.85 ( 0.20% of base) : 231040.dasm - Xunit.Sdk.AssertEqualityComparer`1[Int32][System.Int32]:Equals(int,int):bool:this
        0.85 ( 0.19% of base) : 231055.dasm - Xunit.Sdk.AssertEqualityComparer`1[Int64][System.Int64]:Equals(long,long):bool:this
        0.55 ( 0.04% of base) : 74042.dasm - Microsoft.CodeAnalysis.CommonCompiler:RunCore(System.IO.TextWriter,Microsoft.CodeAnalysis.ErrorLogger,System.Threading.CancellationToken):int:this
        0.50 ( 0.12% of base) : 231835.dasm - Xunit.Sdk.AssertEqualityComparer`1[Byte][System.Byte]:Equals(ubyte,ubyte):bool:this
        0.50 ( 0.12% of base) : 230422.dasm - Xunit.Sdk.AssertEqualityComparer`1[Byte][System.Byte]:Equals(ubyte,ubyte):bool:this
        0.50 ( 0.12% of base) : 231026.dasm - Xunit.Sdk.AssertEqualityComparer`1[Byte][System.Byte]:Equals(ubyte,ubyte):bool:this

Top method improvements (PerfScoreUnits):
       -0.90 (-0.40% of base) : 193548.dasm - <SkipWhileIterator>d__198`1[__Canon][System.__Canon]:MoveNext():bool:this
       -0.75 (-0.20% of base) : 106063.dasm - System.Data.DataColumn:set_DataType(System.Type):this
       -0.48 (-0.59% of base) : 193493.dasm - <SelectIterator>d__180`2[__Canon,Nullable`1][System.__Canon,System.Nullable`1[System.Int32]]:MoveNext():bool:this
       -0.47 (-0.54% of base) : 193611.dasm - <TakeWhileIterator>d__226`1[__Canon][System.__Canon]:MoveNext():bool:this
       -0.47 (-0.20% of base) : 193509.dasm - <SelectManyIterator>d__183`2[__Canon,Nullable`1][System.__Canon,System.Nullable`1[System.Int32]]:MoveNext():bool:this
       -0.47 (-0.21% of base) : 193518.dasm - <SelectManyIterator>d__183`2[Byte,Nullable`1][System.Byte,System.Nullable`1[System.Int32]]:MoveNext():bool:this
       -0.47 (-0.32% of base) : 193632.dasm - <WhereIterator>d__246`1[__Canon][System.__Canon]:MoveNext():bool:this

Top method regressions (percentages):
       32.40 ( 0.63% of base) : 191062.dasm - System.Linq.Parallel.OrderPreservingPipeliningSpoolingTask`2[__Canon,Nullable`1][System.__Canon,System.Nullable`1[System.Int32]]:SpoolingWork():this
        0.90 ( 0.27% of base) : 74188.dasm - Microsoft.CodeAnalysis.AssemblyPortabilityPolicy:LoadFromXml(System.IO.Stream):Microsoft.CodeAnalysis.AssemblyPortabilityPolicy
        0.85 ( 0.20% of base) : 231849.dasm - Xunit.Sdk.AssertEqualityComparer`1[Int32][System.Int32]:Equals(int,int):bool:this
        0.85 ( 0.20% of base) : 230436.dasm - Xunit.Sdk.AssertEqualityComparer`1[Int32][System.Int32]:Equals(int,int):bool:this
        0.85 ( 0.20% of base) : 231040.dasm - Xunit.Sdk.AssertEqualityComparer`1[Int32][System.Int32]:Equals(int,int):bool:this
        0.85 ( 0.19% of base) : 230451.dasm - Xunit.Sdk.AssertEqualityComparer`1[Int64][System.Int64]:Equals(long,long):bool:this
        0.85 ( 0.19% of base) : 231864.dasm - Xunit.Sdk.AssertEqualityComparer`1[Int64][System.Int64]:Equals(long,long):bool:this
        0.85 ( 0.19% of base) : 231055.dasm - Xunit.Sdk.AssertEqualityComparer`1[Int64][System.Int64]:Equals(long,long):bool:this
        0.50 ( 0.12% of base) : 231835.dasm - Xunit.Sdk.AssertEqualityComparer`1[Byte][System.Byte]:Equals(ubyte,ubyte):bool:this
        0.50 ( 0.12% of base) : 230422.dasm - Xunit.Sdk.AssertEqualityComparer`1[Byte][System.Byte]:Equals(ubyte,ubyte):bool:this
        0.50 ( 0.12% of base) : 231026.dasm - Xunit.Sdk.AssertEqualityComparer`1[Byte][System.Byte]:Equals(ubyte,ubyte):bool:this
        0.55 ( 0.04% of base) : 74042.dasm - Microsoft.CodeAnalysis.CommonCompiler:RunCore(System.IO.TextWriter,Microsoft.CodeAnalysis.ErrorLogger,System.Threading.CancellationToken):int:this

Top method improvements (percentages):
       -0.48 (-0.59% of base) : 193493.dasm - <SelectIterator>d__180`2[__Canon,Nullable`1][System.__Canon,System.Nullable`1[System.Int32]]:MoveNext():bool:this
       -0.47 (-0.54% of base) : 193611.dasm - <TakeWhileIterator>d__226`1[__Canon][System.__Canon]:MoveNext():bool:this
       -0.90 (-0.40% of base) : 193548.dasm - <SkipWhileIterator>d__198`1[__Canon][System.__Canon]:MoveNext():bool:this
       -0.47 (-0.32% of base) : 193632.dasm - <WhereIterator>d__246`1[__Canon][System.__Canon]:MoveNext():bool:this
       -0.47 (-0.21% of base) : 193518.dasm - <SelectManyIterator>d__183`2[Byte,Nullable`1][System.Byte,System.Nullable`1[System.Int32]]:MoveNext():bool:this
       -0.75 (-0.20% of base) : 106063.dasm - System.Data.DataColumn:set_DataType(System.Type):this
       -0.47 (-0.20% of base) : 193509.dasm - <SelectManyIterator>d__183`2[__Canon,Nullable`1][System.__Canon,System.Nullable`1[System.Int32]]:MoveNext():bool:this

19 total methods with Perf Score differences (7 improved, 12 regressed), 34 unchanged.

```
</details>

However, I don't see any diffs in libraries if I remove the check for `BBF_SPECIAL_THROW` flag. So I will return `nullptr` for throw block that do not have predecessor.